### PR TITLE
Add DNS ringpop discovery provider

### PR DESCRIPTION
**Pull request by dlh**
Add a new discovery provider to simplify deployments for DNS-based service
discovery platforms, such as HashiCorp Consul and Kubernetes (headless
services).

### CODE REVIEW

File: common/service/config/ringpop.go [hunk 1]
```diff
@@ -202,6 +209,42 @@ func (factory *RingpopFactory) getChannel(dispatcher *yarpc.Dispatcher) (*tcg.Ch
 	return ch, nil
 }
 
+type minimalResolver interface {
+	LookupHost(ctx context.Context, host string) (addrs []string, err error)
+}
+
+type dnsProvider struct {
+	Resolver        minimalResolver
+	UnresolvedHosts []string
```
Comment (venkat1109) [inline common/service/config/ringpop.go]: can you make this unresolved a set ? that way, we can also be sure that there are no duplicates in the input list and so, no duplicates on the output list of resolved addrs. The population of set can be done in a constructor of form newDNSProvider(hosts []string, resolver dnsHostResolver) *dnsProvider {}

File: common/service/config/ringpop.go [hunk 2]
```diff
@@ -202,6 +209,42 @@ func (factory *RingpopFactory) getChannel(dispatcher *yarpc.Dispatcher) (*tcg.Ch
...
+		if !exists {
+			resolved, err = provider.Resolver.LookupHost(context.Background(), host)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		for _, r := range resolved {
+			if !exists {
+				resolvedHosts[host] = append(resolvedHosts[host], r)
```
Comment (venkat1109) [inline common/service/config/ringpop.go]: can you move this code inside of your !exists block before this loop ?

**Comment by dlh** [comment]
@venkat1109 I pushed an update to address your review

File: common/service/config/ringpop.go [hunk 3]
```diff
@@ -202,7 +210,65 @@ func (factory *RingpopFactory) getChannel(dispatcher *yarpc.Dispatcher) (*tcg.Ch
...
+		}
+
+		resolved, exists := resolvedHosts[host]
+		if !exists {
+			resolved, err = provider.Resolver.LookupHost(context.Background(), host)
+			if err != nil {
+				provider.Logger.Warn("Could not resolve host", tag.Address(host), tag.Error(err))
+				continue
+			}
+			for _, r := range resolved {
```
Comment (venkat1109) [inline common/service/config/ringpop.go]: are there reasons to do a slice copy here instead of simply doing  ```resolvedHosts[host] = resolved``` ?

Comment (dlh) [reply common/service/config/ringpop.go]: 🤦‍♂ Nope no reason. I updated it to be direct assignment.

**Pull request merged**